# Pandas Assignment: 50 Basic Questions

Dataset Columns: CID, TID, Gender, Age Group, Purchase Date, Product Category, Discount Availed, Discount Name, Discount Amount (INR), Gross Amount, Net Amount, Purchase Method, Location

### 1. Load the CSV file into a Pandas DataFrame.

In [1]:
import pandas as pd

In [76]:
df = pd.read_csv("data.csv")
print(df)

          CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
3      180333  1395204173   Other  60 and above  04/11/2020 04:41:57   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            Electronics              Yes        FESTIVE50   
1          

### 2. Display the first 5 rows.

In [10]:
df.head()

CID         TID  Gender     Age Group        Purchase Date  \
0  943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1  180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2  337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
3  180333  1395204173   Other  60 and above  04/11/2020 04:41:57   
4  447553  8009390577    Male         18-25  31/05/2022 17:00:32   

   Product Category Discount Availed    Discount Name  Discount Amount (INR)  \
0       Electronics              Yes        FESTIVE50                  64.30   
1       Electronics              Yes  SEASONALOFFER21                 175.19   
2          Clothing              Yes  SEASONALOFFER21                 211.54   
3  Sports & Fitness               No              NaN                   0.00   
4  Sports & Fitness              Yes         WELCOME5                 439.92   

   Gross Amount   Net Amount Purchase Method   Location  
0    725.304000   661.004000     Credit Card  Ahmedabad  
1   4638.991875  4463.801875     Credit Card  Bangalore  
2   1986.372575  1774.832575     Credit Card      Delhi  
3   5695.612650  5695.612650      Debit Card      Delhi  
4   2292.651500  1852.731500     Credit Card      Delhi

In [ ]:
# df.head().columns   to check column name

Index(['CID', 'TID', 'Gender', 'Age Group', 'Purchase Date',
       'Product Category', 'Discount Availed', 'Discount Name',
       'Discount Amount (INR)', 'Gross Amount', 'Net Amount',
       'Purchase Method', 'Location'],
      dtype='object')

### 3. Display the last 5 rows.

In [12]:
df.tail()

CID         TID  Gender Age Group        Purchase Date  \
54995  295864  5407639888   Other     25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other     18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male     45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female     45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male     25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
54995        Electronics              Yes  SEASONALOFFER21   
54996        Electronics              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount  Net Amount   Purchase Method  \
54995                 110.89     2819.9073   2709.0173        Debit Card   
54996                 488.59     3891.3039   3402.7139  Cash on Delivery   
54997                 417.63     4094.6875   3677.0575       Credit Card   
54998                 204.56     3977.4735   3772.9135       Credit Card   
54999                 380.81     3703.5819   3322.7719       Credit Card   

        Location  
54995      Other  
54996  Bangalore  
54997     Mumbai  
54998    Kolkata  
54999     Mumbai

### 4. Print the shape of the DataFrame.

In [21]:
df.shape
# df.shape[0]  # rows only
# df.shape[1]  # column 

(55000, 13)

### 5. Show all column names.

In [22]:
df.columns

Index(['CID', 'TID', 'Gender', 'Age Group', 'Purchase Date',
       'Product Category', 'Discount Availed', 'Discount Name',
       'Discount Amount (INR)', 'Gross Amount', 'Net Amount',
       'Purchase Method', 'Location'],
      dtype='object')

### 6. Display the data types of each column.

In [24]:
df.dtypes

CID                        int64
TID                        int64
Gender                    object
Age Group                 object
Purchase Date             object
Product Category          object
Discount Availed          object
Discount Name             object
Discount Amount (INR)    float64
Gross Amount             float64
Net Amount               float64
Purchase Method           object
Location                  object
dtype: object

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CID                    55000 non-null  int64  
 1   TID                    55000 non-null  int64  
 2   Gender                 55000 non-null  object 
 3   Age Group              55000 non-null  object 
 4   Purchase Date          55000 non-null  object 
 5   Product Category       55000 non-null  object 
 6   Discount Availed       55000 non-null  object 
 7   Discount Name          27415 non-null  object 
 8   Discount Amount (INR)  55000 non-null  float64
 9   Gross Amount           55000 non-null  float64
 10  Net Amount             55000 non-null  float64
 11  Purchase Method        55000 non-null  object 
 12  Location               55000 non-null  object 
dtypes: float64(3), int64(2), object(8)
memory usage: 5.5+ MB


### 7. Check for missing values in each column.

In [26]:
df.isna().sum()

CID                          0
TID                          0
Gender                       0
Age Group                    0
Purchase Date                0
Product Category             0
Discount Availed             0
Discount Name            27585
Discount Amount (INR)        0
Gross Amount                 0
Net Amount                   0
Purchase Method              0
Location                     0
dtype: int64

In [25]:
df.isnull().sum()

CID                          0
TID                          0
Gender                       0
Age Group                    0
Purchase Date                0
Product Category             0
Discount Availed             0
Discount Name            27585
Discount Amount (INR)        0
Gross Amount                 0
Net Amount                   0
Purchase Method              0
Location                     0
dtype: int64

### 8. Count total missing values in the dataset.

In [29]:
total_missing = int(df.isnull().sum().sum())
total_missing

27585

In [28]:
df.isna().sum().sum()

np.int64(27585)

### 9. Show summary statistics for numeric columns.

In [30]:
df.describe()

CID           TID  Discount Amount (INR)  Gross Amount  \
count   55000.000000  5.500000e+04           55000.000000  55000.000000   
mean   551245.593891  5.504740e+09             136.986796   3012.936606   
std    260603.330337  2.594534e+09             165.375502   1718.431066   
min    100009.000000  1.000163e+09               0.000000    136.454325   
25%    323717.000000  3.252604e+09               0.000000   1562.111325   
50%    550088.500000  5.498383e+09               0.000000   2954.266150   
75%    776955.750000  7.747933e+09             274.115000   4342.221675   
max    999996.000000  9.999393e+09             500.000000   8394.825600   

         Net Amount  
count  55000.000000  
mean    2875.949810  
std     1726.127778  
min     -351.119775  
25%     1429.551863  
50%     2814.910875  
75%     4211.407838  
max     8394.825600

### 10. Display info about the DataFrame using .info().

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CID                    55000 non-null  int64  
 1   TID                    55000 non-null  int64  
 2   Gender                 55000 non-null  object 
 3   Age Group              55000 non-null  object 
 4   Purchase Date          55000 non-null  object 
 5   Product Category       55000 non-null  object 
 6   Discount Availed       55000 non-null  object 
 7   Discount Name          27415 non-null  object 
 8   Discount Amount (INR)  55000 non-null  float64
 9   Gross Amount           55000 non-null  float64
 10  Net Amount             55000 non-null  float64
 11  Purchase Method        55000 non-null  object 
 12  Location               55000 non-null  object 
dtypes: float64(3), int64(2), object(8)
memory usage: 5.5+ MB


### 11. Select the CID and TID columns.

In [32]:
df[['CID','TID']]

CID         TID
0      943146  5876328741
1      180079  1018503182
2      337580  3814082218
3      180333  1395204173
4      447553  8009390577
...       ...         ...
54995  295864  5407639888
54996  706435  1144029904
54997  649435  9587323189
54998  982183  5032564481
54999  126691  9172953722

[55000 rows x 2 columns]

### 12. Display the first 10 rows of Gender and Age Group.

In [6]:
df[["Gender","Age Group"]].head(10)

Gender     Age Group
0  Female         25-45
1    Male         25-45
2   Other  60 and above
3   Other  60 and above
4    Male         18-25
5    Male         18-25
6    Male         25-45
7  Female         18-25
8    Male         18-25
9   Other  60 and above

### 13. Filter rows where Gender is 'Male'.

In [ ]:
df[df["Gender"] == "Male"]

CID         TID Gender Age Group        Purchase Date  \
1      180079  1018503182   Male     25-45  23/02/2024 09:33:46   
4      447553  8009390577   Male     18-25  31/05/2022 17:00:32   
5      200614  3994452858   Male     18-25  12/07/2021 15:10:27   
6      668787  9460023916   Male     25-45  08/10/2023 16:59:52   
8      414167  2841266620   Male     18-25  29/04/2020 09:10:46   
...       ...         ...    ...       ...                  ...   
54989  812597  3862962312   Male     45-60  07/08/2022 11:42:13   
54990  180783  4001403523   Male     25-45  06/08/2024 23:15:21   
54992  639165  8525877571   Male     25-45  15/05/2022 20:32:26   
54997  649435  9587323189   Male     45-60  02/08/2024 09:30:44   
54999  126691  9172953722   Male     25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
1            Electronics              Yes  SEASONALOFFER21   
4       Sports & Fitness              Yes         WELCOME5   
5               Clothing              Yes        FESTIVE50   
6            Electronics               No              NaN   
8               Pet Care               No              NaN   
...                  ...              ...              ...   
54989           Clothing               No              NaN   
54990     Home & Kitchen               No              NaN   
54992        Electronics               No              NaN   
54997     Home & Kitchen              Yes         NEWYEARS   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount Purchase Method  \
1                     175.19   4638.991875  4463.801875     Credit Card   
4                     439.92   2292.651500  1852.731500     Credit Card   
5                     127.01   3649.397400  3522.387400     Credit Card   
6                       0.00   5593.935600  5593.935600     PhonePe UPI   
8                       0.00   3961.671000  3961.671000      Debit Card   
...                      ...           ...          ...             ...   
54989                   0.00   2335.098000  2335.098000     Credit Card   
54990                   0.00   2679.337500  2679.337500     Credit Card   
54992                   0.00   4489.956500  4489.956500     Credit Card   
54997                 417.63   4094.687500  3677.057500     Credit Card   
54999                 380.81   3703.581900  3322.771900     Credit Card   

        Location  
1      Bangalore  
4          Delhi  
5          Delhi  
6          Delhi  
8          Delhi  
...          ...  
54989       Pune  
54990      Delhi  
54992    Kolkata  
54997     Mumbai  
54999     Mumbai  

[18096 rows x 13 columns]

### 14. Filter rows where Age Group is '26-35'.

In [13]:
df[df["Age Group"] == "26-35" ]

Empty DataFrame
Columns: [CID, TID, Gender, Age Group, Purchase Date, Product Category, Discount Availed, Discount Name, Discount Amount (INR), Gross Amount, Net Amount, Purchase Method, Location]
Index: []

### 15. Filter rows where Discount Availed is 'Yes'.

In [14]:
df[df["Discount Availed"] == "Yes"]

CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
5      200614  3994452858    Male         18-25  12/07/2021 15:10:27   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            Electronics              Yes        FESTIVE50   
1            Electronics              Yes  SEASONALOFFER21   
2               Clothing              Yes  SEASONALOFFER21   
4       Sports & Fitness              Yes         WELCOME5   
5               Clothing              Yes        FESTIVE50   
...                  ...              ...              ...   
54995        Electronics              Yes  SEASONALOFFER21   
54996        Electronics              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
0                      64.30    725.304000   661.004000       Credit Card   
1                     175.19   4638.991875  4463.801875       Credit Card   
2                     211.54   1986.372575  1774.832575       Credit Card   
4                     439.92   2292.651500  1852.731500       Credit Card   
5                     127.01   3649.397400  3522.387400       Credit Card   
...                      ...           ...          ...               ...   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   
54997                 417.63   4094.687500  3677.057500       Credit Card   
54998                 204.56   3977.473500  3772.913500       Credit Card   
54999                 380.81   3703.581900  3322.771900       Credit Card   

        Location  
0      Ahmedabad  
1      Bangalore  
2          Delhi  
4          Delhi  
5          Delhi  
...          ...  
54995      Other  
54996  Bangalore  
54997     Mumbai  
54998    Kolkata  
54999     Mumbai  

[27415 rows x 13 columns]

### 16. Filter rows where Discount Amount (INR) > 500.

In [19]:
df[df["Discount Amount (INR)"] > 500]

Empty DataFrame
Columns: [CID, TID, Gender, Age Group, Purchase Date, Product Category, Discount Availed, Discount Name, Discount Amount (INR), Gross Amount, Net Amount, Purchase Method, Location]
Index: []

### 17. Filter rows where Product Category is 'Electronics'.

In [20]:
df[df["Product Category"]=="Electronics"]

CID         TID  Gender Age Group        Purchase Date  \
0      943146  5876328741  Female     25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male     25-45  23/02/2024 09:33:46   
6      668787  9460023916    Male     25-45  08/10/2023 16:59:52   
7      933551  6877673032  Female     18-25  14/12/2022 20:13:45   
11     629298  2229511931   Other     45-60  02/11/2021 15:37:38   
...       ...         ...     ...       ...                  ...   
54985  903683  2516034122  Female     25-45  22/03/2024 23:52:20   
54987  907365  2841576857    Male     25-45  12/05/2023 15:48:26   
54992  639165  8525877571    Male     25-45  15/05/2022 20:32:26   
54995  295864  5407639888   Other     25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other     18-25  12/07/2021 15:59:21   

      Product Category Discount Availed    Discount Name  \
0          Electronics              Yes        FESTIVE50   
1          Electronics              Yes  SEASONALOFFER21   
6          Electronics               No              NaN   
7          Electronics               No              NaN   
11         Electronics               No              NaN   
...                ...              ...              ...   
54985      Electronics              Yes         NEWYEARS   
54987      Electronics               No              NaN   
54992      Electronics               No              NaN   
54995      Electronics              Yes  SEASONALOFFER21   
54996      Electronics              Yes  SEASONALOFFER21   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
0                      64.30    725.304000   661.004000       Credit Card   
1                     175.19   4638.991875  4463.801875       Credit Card   
6                       0.00   5593.935600  5593.935600       PhonePe UPI   
7                       0.00   1901.796400  1901.796400       Credit Card   
11                      0.00    988.172900   988.172900    Google Pay UPI   
...                      ...           ...          ...               ...   
54985                 456.01   4733.826250  4277.816250         Paytm UPI   
54987                   0.00    927.384000   927.384000        Debit Card   
54992                   0.00   4489.956500  4489.956500       Credit Card   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   

        Location  
0      Ahmedabad  
1      Bangalore  
6          Delhi  
7          Delhi  
11       Chennai  
...          ...  
54985     Mumbai  
54987  Bangalore  
54992    Kolkata  
54995      Other  
54996  Bangalore  

[16574 rows x 13 columns]

### 18. Display rows where Purchase Method is 'Credit Card'.

In [21]:
df[df["Purchase Method"] == "Credit Card"]

CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
5      200614  3994452858    Male         18-25  12/07/2021 15:10:27   
...       ...         ...     ...           ...                  ...   
54991  103234  5584476454   Other         45-60  29/08/2021 18:47:25   
54992  639165  8525877571    Male         25-45  15/05/2022 20:32:26   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            Electronics              Yes        FESTIVE50   
1            Electronics              Yes  SEASONALOFFER21   
2               Clothing              Yes  SEASONALOFFER21   
4       Sports & Fitness              Yes         WELCOME5   
5               Clothing              Yes        FESTIVE50   
...                  ...              ...              ...   
54991              Other               No              NaN   
54992        Electronics               No              NaN   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount Purchase Method  \
0                      64.30    725.304000   661.004000     Credit Card   
1                     175.19   4638.991875  4463.801875     Credit Card   
2                     211.54   1986.372575  1774.832575     Credit Card   
4                     439.92   2292.651500  1852.731500     Credit Card   
5                     127.01   3649.397400  3522.387400     Credit Card   
...                      ...           ...          ...             ...   
54991                   0.00   2385.438000  2385.438000     Credit Card   
54992                   0.00   4489.956500  4489.956500     Credit Card   
54997                 417.63   4094.687500  3677.057500     Credit Card   
54998                 204.56   3977.473500  3772.913500     Credit Card   
54999                 380.81   3703.581900  3322.771900     Credit Card   

        Location  
0      Ahmedabad  
1      Bangalore  
2          Delhi  
4          Delhi  
5          Delhi  
...          ...  
54991  Bangalore  
54992    Kolkata  
54997     Mumbai  
54998    Kolkata  
54999     Mumbai  

[22096 rows x 13 columns]

### 19. Select rows where Location is 'Mumbai'.

In [22]:
df[df["Location"] == "Mumbai"]

CID         TID  Gender     Age Group        Purchase Date  \
19     787512  9582806381    Male         18-25  17/03/2021 17:32:46   
20     538174  1322344974    Male  60 and above  24/11/2020 02:42:48   
22     443120  4473452038  Female      under 18  07/09/2023 23:23:03   
24     745014  9202622469    Male         18-25  04/05/2023 18:06:02   
26     654718  6758404906   Other      under 18  22/07/2023 23:26:04   
...       ...         ...     ...           ...                  ...   
54983  306954  1846818218  Female         18-25  13/09/2023 06:55:26   
54985  903683  2516034122  Female         25-45  22/03/2024 23:52:20   
54986  915197  5506311767  Female         25-45  06/03/2023 15:57:05   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
19                 Books               No              NaN   
20      Sports & Fitness               No              NaN   
22      Sports & Fitness              Yes         NEWYEARS   
24              Clothing              Yes           SAVE10   
26     Beauty and Health               No              NaN   
...                  ...              ...              ...   
54983   Sports & Fitness               No              NaN   
54985        Electronics              Yes         NEWYEARS   
54986           Clothing              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount  Net Amount   Purchase Method  \
19                      0.00    1952.41530  1952.41530         Paytm UPI   
20                      0.00    5025.71160  5025.71160       Net Banking   
22                    268.54    1495.20420  1226.66420        Debit Card   
24                     75.72    3579.91200  3504.19200       Credit Card   
26                      0.00    5057.99640  5057.99640        Debit Card   
...                      ...           ...         ...               ...   
54983                   0.00     892.99980   892.99980       PhonePe UPI   
54985                 456.01    4733.82625  4277.81625         Paytm UPI   
54986                 489.26    2425.12200  1935.86200  Cash on Delivery   
54997                 417.63    4094.68750  3677.05750       Credit Card   
54999                 380.81    3703.58190  3322.77190       Credit Card   

      Location  
19      Mumbai  
20      Mumbai  
22      Mumbai  
24      Mumbai  
26      Mumbai  
...        ...  
54983   Mumbai  
54985   Mumbai  
54986   Mumbai  
54997   Mumbai  
54999   Mumbai  

[11197 rows x 13 columns]

### 20. Select rows where Gross Amount is between 1000 and 5000.

In [23]:
df[(df["Gross Amount"] >= 1000) & (df["Gross Amount"] <= 5000)]

CID         TID  Gender     Age Group        Purchase Date  \
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
5      200614  3994452858    Male         18-25  12/07/2021 15:10:27   
7      933551  6877673032  Female         18-25  14/12/2022 20:13:45   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
1            Electronics              Yes  SEASONALOFFER21   
2               Clothing              Yes  SEASONALOFFER21   
4       Sports & Fitness              Yes         WELCOME5   
5               Clothing              Yes        FESTIVE50   
7            Electronics               No              NaN   
...                  ...              ...              ...   
54995        Electronics              Yes  SEASONALOFFER21   
54996        Electronics              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
1                     175.19   4638.991875  4463.801875       Credit Card   
2                     211.54   1986.372575  1774.832575       Credit Card   
4                     439.92   2292.651500  1852.731500       Credit Card   
5                     127.01   3649.397400  3522.387400       Credit Card   
7                       0.00   1901.796400  1901.796400       Credit Card   
...                      ...           ...          ...               ...   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   
54997                 417.63   4094.687500  3677.057500       Credit Card   
54998                 204.56   3977.473500  3772.913500       Credit Card   
54999                 380.81   3703.581900  3322.771900       Credit Card   

        Location  
1      Bangalore  
2          Delhi  
4          Delhi  
5          Delhi  
7          Delhi  
...          ...  
54995      Other  
54996  Bangalore  
54997     Mumbai  
54998    Kolkata  
54999     Mumbai  

[39140 rows x 13 columns]

### 21. Sort the DataFrame by Gross Amount ascending.

In [33]:
df.sort_values(by = "Gross Amount",ascending = True)

CID         TID  Gender Age Group        Purchase Date  \
37352  570046  1964089589  Female     25-45  18/02/2020 02:14:13   
17568  262070  5537538360  Female  under 18  01/02/2020 00:15:47   
38606  481252  9891577351  Female     25-45  03/02/2020 07:18:59   
40677  376537  8659700479  Female     25-45  14/01/2020 19:49:31   
48164  386840  8029271119  Female     18-25  20/02/2020 14:18:22   
...       ...         ...     ...       ...                  ...   
52533  675956  7090926391  Female  under 18  01/12/2023 07:55:47   
5521   255607  7392860072    Male     18-25  15/12/2023 15:55:16   
29249  871976  7829441574  Female     25-45  19/12/2023 09:49:50   
21848  738513  7204773289  Female     18-25  06/12/2023 04:26:33   
26874  202767  5202191880   Other     18-25  11/12/2023 11:19:48   

        Product Category Discount Availed    Discount Name  \
37352  Beauty and Health               No              NaN   
17568        Electronics              Yes         NEWYEARS   
38606           Clothing              Yes  SEASONALOFFER21   
40677           Clothing               No              NaN   
48164        Electronics              Yes         NEWYEARS   
...                  ...              ...              ...   
52533        Electronics              Yes  SEASONALOFFER21   
5521            Clothing               No              NaN   
29249       Toys & Games              Yes        FESTIVE50   
21848  Beauty and Health               No              NaN   
26874     Home & Kitchen               No              NaN   

       Discount Amount (INR)  Gross Amount   Net Amount Purchase Method  \
37352                   0.00    136.454325   136.454325     Credit Card   
17568                 429.60    138.890850  -290.709150     Credit Card   
38606                 413.14    140.461650  -272.678350  Google Pay UPI   
40677                   0.00    141.957900   141.957900  Google Pay UPI   
48164                 352.96    142.077075  -210.882925     Credit Card   
...                      ...           ...          ...             ...   
52533                 214.33   8334.009600  8119.679600      Debit Card   
5521                    0.00   8361.225600  8361.225600     Credit Card   
29249                 492.56   8365.190400  7872.630400     Net Banking   
21848                   0.00   8382.796800  8382.796800      Debit Card   
26874                   0.00   8394.825600  8394.825600      Debit Card   

        Location  
37352   Varanasi  
17568     Mumbai  
38606  Hyderabad  
40677   Varanasi  
48164  Ahmedabad  
...          ...  
52533    Chennai  
5521      Mumbai  
29249      Delhi  
21848      Delhi  
26874       Pune  

[55000 rows x 13 columns]

### 22. Sort the DataFrame by Net Amount descending.

In [29]:
df.sort_values(by="Net Amount", ascending = False)

CID         TID  Gender Age Group        Purchase Date  \
26874  202767  5202191880   Other     18-25  11/12/2023 11:19:48   
21848  738513  7204773289  Female     18-25  06/12/2023 04:26:33   
5521   255607  7392860072    Male     18-25  15/12/2023 15:55:16   
46212  327575  7133917620   Other  under 18  02/12/2023 07:47:56   
27657  433217  1340357202    Male     45-60  21/12/2023 12:16:26   
...       ...         ...     ...       ...                  ...   
16031  574966  9576162844  Female     25-45  30/01/2023 00:53:06   
2606   472724  2853800882   Other     18-25  03/04/2020 16:08:07   
22138  761286  3119749374    Male     45-60  15/06/2022 12:08:59   
7152   717101  8220478366    Male     25-45  10/02/2022 22:38:06   
18827  898531  4861274119   Other     45-60  23/02/2020 05:26:39   

        Product Category Discount Availed    Discount Name  \
26874     Home & Kitchen               No              NaN   
21848  Beauty and Health               No              NaN   
5521            Clothing               No              NaN   
46212           Clothing               No              NaN   
27657        Electronics               No              NaN   
...                  ...              ...              ...   
16031        Electronics              Yes  SEASONALOFFER21   
2606            Clothing              Yes  SEASONALOFFER21   
22138           Clothing              Yes         NEWYEARS   
7152         Electronics              Yes           SAVE10   
18827        Electronics              Yes  SEASONALOFFER21   

       Discount Amount (INR)  Gross Amount   Net Amount Purchase Method  \
26874                   0.00   8394.825600  8394.825600      Debit Card   
21848                   0.00   8382.796800  8382.796800      Debit Card   
5521                    0.00   8361.225600  8361.225600     Credit Card   
46212                   0.00   8309.767200  8309.767200     Credit Card   
27657                   0.00   8238.535200  8238.535200     Credit Card   
...                      ...           ...          ...             ...   
16031                 482.13    171.871200  -310.258800     Credit Card   
2606                  495.73    182.731500  -312.998500     Credit Card   
22138                 485.12    159.089850  -326.030150     Credit Card   
7152                  491.35    164.757625  -326.592375     Credit Card   
18827                 494.25    143.130225  -351.119775     Credit Card   

        Location  
26874       Pune  
21848      Delhi  
5521      Mumbai  
46212  Ahmedabad  
27657  Bangalore  
...          ...  
16031  Ahmedabad  
2606   Bangalore  
22138       Pune  
7152      Mumbai  
18827  Bangalore  

[55000 rows x 13 columns]

### 23. Sort by Location and then by Purchase Date.

In [36]:
df.sort_values(
    by=["Location","Purchase Date"],
    ascending=True
)

CID         TID  Gender     Age Group        Purchase Date  \
14677  356972  9481824682  Female         25-45  01/01/2020 16:36:35   
52154  291523  3456432695   Other         25-45  01/01/2020 19:55:04   
36766  636294  8503669581  Female         25-45  01/01/2021 10:36:31   
47676  582716  7455858410   Other         25-45  01/01/2022 03:01:35   
26695  380678  4378851841    Male         25-45  01/01/2022 07:47:24   
...       ...         ...     ...           ...                  ...   
8966   298020  1127367408    Male         25-45  31/08/2023 16:58:41   
4011   502349  3964650442  Female         18-25  31/08/2024 03:59:17   
24671  374505  8692673674  Female         25-45  31/08/2024 11:51:32   
50120  486244  2152632978  Female         25-45  31/10/2019 13:34:58   
48865  502580  6251204383    Male  60 and above  31/12/2023 09:50:47   

        Product Category Discount Availed    Discount Name  \
14677              Books              Yes           SAVE10   
52154   Sports & Fitness               No              NaN   
36766       Toys & Games              Yes         NEWYEARS   
47676           Clothing               No              NaN   
26695           Clothing              Yes  SEASONALOFFER21   
...                  ...              ...              ...   
8966         Electronics              Yes         NEWYEARS   
4011   Beauty and Health              Yes  SEASONALOFFER21   
24671              Other              Yes           SAVE10   
50120              Other               No              NaN   
48865           Clothing              Yes  SEASONALOFFER21   

       Discount Amount (INR)  Gross Amount  Net Amount     Purchase Method  \
14677                  81.76     4201.6968   4119.9368          Debit Card   
52154                   0.00     1121.9418   1121.9418    Cash on Delivery   
36766                 457.49     3911.0544   3453.5644  International Card   
47676                   0.00     4558.0779   4558.0779           Paytm UPI   
26695                 138.82     1121.6502    982.8302         Credit Card   
...                      ...           ...         ...                 ...   
8966                  294.44     5206.6440   4912.2040          Debit Card   
4011                  383.84     1208.3375    824.4975         Credit Card   
24671                 316.20     3370.3000   3054.1000         Credit Card   
50120                   0.00     2197.7120   2197.7120         Credit Card   
48865                  57.98     4926.7512   4868.7712          Debit Card   

        Location  
14677  Ahmedabad  
52154  Ahmedabad  
36766  Ahmedabad  
47676  Ahmedabad  
26695  Ahmedabad  
...          ...  
8966    Varanasi  
4011    Varanasi  
24671   Varanasi  
50120   Varanasi  
48865   Varanasi  

[55000 rows x 13 columns]

### 24. Find the number of unique locations.

In [40]:
df["Location"].nunique()

14

In [ ]:
df["Location"].value_counts()      # unique value with freq

Location
Mumbai       11197
Delhi        10799
Bangalore     8249
Hyderabad     5545
Chennai       4368
Pune          3781
Ahmedabad     2785
Kolkata       2709
Jaipur        1678
Lucknow       1136
Other         1046
Varanasi       606
Dehradun       553
Srinagar       548
Name: count, dtype: int64

 ### 25. Display all unique locations. 

In [41]:
df["Location"].unique()

array(['Ahmedabad', 'Bangalore', 'Delhi', 'Other', 'Chennai', 'Dehradun',
       'Pune', 'Hyderabad', 'Mumbai', 'Jaipur', 'Lucknow', 'Kolkata',
       'Srinagar', 'Varanasi'], dtype=object)

### 26. Find the number of unique product categories.

In [46]:
df["Product Category"].nunique()

9

### 27. Display the top 10 most frequent product categories.

In [50]:
df["Product Category"].value_counts().head(10)

Product Category
Electronics          16574
Clothing             10968
Beauty and Health     8332
Sports & Fitness      5557
Home & Kitchen        5489
Books                 2762
Other                 2171
Pet Care              1618
Toys & Games          1529
Name: count, dtype: int64

### 28. Display the top 10 most frequent locations.

In [51]:
df["Location"].value_counts().head(10)

Location
Mumbai       11197
Delhi        10799
Bangalore     8249
Hyderabad     5545
Chennai       4368
Pune          3781
Ahmedabad     2785
Kolkata       2709
Jaipur        1678
Lucknow       1136
Name: count, dtype: int64

### 29. Find the transaction with the highest Gross Amount.

In [71]:
df[df["Gross Amount"] == df["Gross Amount"].max()][["TID","Gross Amount"]]

TID  Gross Amount
26874  5202191880     8394.8256

### 30. Find the transaction with the highest Discount Amount (INR).

In [73]:
df[df["Discount Amount (INR)"] == df["Discount Amount (INR)"].max()]

CID         TID  Gender Age Group        Purchase Date  \
37549  108739  8375499682  Female     45-60  24/10/2023 09:29:36   

        Product Category Discount Availed    Discount Name  \
37549  Beauty and Health              Yes  SEASONALOFFER21   

       Discount Amount (INR)  Gross Amount  Net Amount Purchase Method  \
37549                  500.0      604.6656    104.6656       Paytm UPI   

      Location  
37549  Chennai

In [74]:
df[df["Discount Amount (INR)"] == df["Discount Amount (INR)"].max()][["TID","Discount Amount (INR)"]]

TID  Discount Amount (INR)
37549  8375499682                  500.0

### 31. Create a new column Discount Percentage = (Discount Amount / Gross Amount) * 100.

In [3]:
df["Discount Percentage"] = (df["Discount Amount (INR)"]/ df["Gross Amount"]) * 100

In [4]:
df

CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
3      180333  1395204173   Other  60 and above  04/11/2020 04:41:57   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            Electronics              Yes        FESTIVE50   
1            Electronics              Yes  SEASONALOFFER21   
2               Clothing              Yes  SEASONALOFFER21   
3       Sports & Fitness               No              NaN   
4       Sports & Fitness              Yes         WELCOME5   
...                  ...              ...              ...   
54995        Electronics              Yes  SEASONALOFFER21   
54996        Electronics              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
0                      64.30    725.304000   661.004000       Credit Card   
1                     175.19   4638.991875  4463.801875       Credit Card   
2                     211.54   1986.372575  1774.832575       Credit Card   
3                       0.00   5695.612650  5695.612650        Debit Card   
4                     439.92   2292.651500  1852.731500       Credit Card   
...                      ...           ...          ...               ...   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   
54997                 417.63   4094.687500  3677.057500       Credit Card   
54998                 204.56   3977.473500  3772.913500       Credit Card   
54999                 380.81   3703.581900  3322.771900       Credit Card   

        Location  Discount Percentage  
0      Ahmedabad             8.865248  
1      Bangalore             3.776467  
2          Delhi            10.649563  
3          Delhi             0.000000  
4          Delhi            19.188263  
...          ...                  ...  
54995      Other             3.932399  
54996  Bangalore            12.555946  
54997     Mumbai            10.199313  
54998    Kolkata             5.142963  
54999     Mumbai            10.282208  

[55000 rows x 14 columns]

### 32. Round Discount Percentage to 2 decimal places.

In [86]:
df["Discount Percentage"] = df["Discount Percentage"].round(2)

In [88]:
# df["Discount Percentage"].round(2) ## Display without changing value

In [87]:
df

CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
3      180333  1395204173   Other  60 and above  04/11/2020 04:41:57   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            Electronics              Yes        FESTIVE50   
1            Electronics              Yes  SEASONALOFFER21   
2               Clothing              Yes  SEASONALOFFER21   
3       Sports & Fitness               No              NaN   
4       Sports & Fitness              Yes         WELCOME5   
...                  ...              ...              ...   
54995        Electronics              Yes  SEASONALOFFER21   
54996        Electronics              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
0                      64.30    725.304000   661.004000       Credit Card   
1                     175.19   4638.991875  4463.801875       Credit Card   
2                     211.54   1986.372575  1774.832575       Credit Card   
3                       0.00   5695.612650  5695.612650        Debit Card   
4                     439.92   2292.651500  1852.731500       Credit Card   
...                      ...           ...          ...               ...   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   
54997                 417.63   4094.687500  3677.057500       Credit Card   
54998                 204.56   3977.473500  3772.913500       Credit Card   
54999                 380.81   3703.581900  3322.771900       Credit Card   

        Location  Discount Percentage  
0      Ahmedabad                 8.87  
1      Bangalore                 3.78  
2          Delhi                10.65  
3          Delhi                 0.00  
4          Delhi                19.19  
...          ...                  ...  
54995      Other                 3.93  
54996  Bangalore                12.56  
54997     Mumbai                10.20  
54998    Kolkata                 5.14  
54999     Mumbai                10.28  

[55000 rows x 14 columns]

### 33. Convert Product Category to uppercase.

In [90]:
df["Product Category"] = df["Product Category"].str.upper()

In [91]:
# df["Product Category"].str.upper()

In [92]:
df

CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
3      180333  1395204173   Other  60 and above  04/11/2020 04:41:57   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            ELECTRONICS              Yes        FESTIVE50   
1            ELECTRONICS              Yes  SEASONALOFFER21   
2               CLOTHING              Yes  SEASONALOFFER21   
3       SPORTS & FITNESS               No              NaN   
4       SPORTS & FITNESS              Yes         WELCOME5   
...                  ...              ...              ...   
54995        ELECTRONICS              Yes  SEASONALOFFER21   
54996        ELECTRONICS              Yes  SEASONALOFFER21   
54997     HOME & KITCHEN              Yes         NEWYEARS   
54998  BEAUTY AND HEALTH              Yes  SEASONALOFFER21   
54999  BEAUTY AND HEALTH              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
0                      64.30    725.304000   661.004000       Credit Card   
1                     175.19   4638.991875  4463.801875       Credit Card   
2                     211.54   1986.372575  1774.832575       Credit Card   
3                       0.00   5695.612650  5695.612650        Debit Card   
4                     439.92   2292.651500  1852.731500       Credit Card   
...                      ...           ...          ...               ...   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   
54997                 417.63   4094.687500  3677.057500       Credit Card   
54998                 204.56   3977.473500  3772.913500       Credit Card   
54999                 380.81   3703.581900  3322.771900       Credit Card   

        Location  Discount Percentage  
0      Ahmedabad                 8.87  
1      Bangalore                 3.78  
2          Delhi                10.65  
3          Delhi                 0.00  
4          Delhi                19.19  
...          ...                  ...  
54995      Other                 3.93  
54996  Bangalore                12.56  
54997     Mumbai                10.20  
54998    Kolkata                 5.14  
54999     Mumbai                10.28  

[55000 rows x 14 columns]

### 34. Remove leading/trailing spaces from Discount Name.

In [94]:
df["Discount Name"]=df["Discount Name"].str.strip()

### 35. Replace missing Discount Name with 'No Discount'.

In [8]:
df["Discount Name"] = df["Discount Name"].fillna("No Discount")

In [11]:
df

CID         TID  Gender     Age Group        Purchase Date  \
0      943146  5876328741  Female         25-45  30/08/2023 20:27:08   
1      180079  1018503182    Male         25-45  23/02/2024 09:33:46   
2      337580  3814082218   Other  60 and above  06/03/2022 09:09:50   
3      180333  1395204173   Other  60 and above  04/11/2020 04:41:57   
4      447553  8009390577    Male         18-25  31/05/2022 17:00:32   
...       ...         ...     ...           ...                  ...   
54995  295864  5407639888   Other         25-45  23/02/2020 10:42:04   
54996  706435  1144029904   Other         18-25  12/07/2021 15:59:21   
54997  649435  9587323189    Male         45-60  02/08/2024 09:30:44   
54998  982183  5032564481  Female         45-60  05/08/2020 23:57:56   
54999  126691  9172953722    Male         25-45  21/07/2022 09:05:18   

        Product Category Discount Availed    Discount Name  \
0            Electronics              Yes        FESTIVE50   
1            Electronics              Yes  SEASONALOFFER21   
2               Clothing              Yes  SEASONALOFFER21   
3       Sports & Fitness               No      No Discount   
4       Sports & Fitness              Yes         WELCOME5   
...                  ...              ...              ...   
54995        Electronics              Yes  SEASONALOFFER21   
54996        Electronics              Yes  SEASONALOFFER21   
54997     Home & Kitchen              Yes         NEWYEARS   
54998  Beauty and Health              Yes  SEASONALOFFER21   
54999  Beauty and Health              Yes         NEWYEARS   

       Discount Amount (INR)  Gross Amount   Net Amount   Purchase Method  \
0                      64.30    725.304000   661.004000       Credit Card   
1                     175.19   4638.991875  4463.801875       Credit Card   
2                     211.54   1986.372575  1774.832575       Credit Card   
3                       0.00   5695.612650  5695.612650        Debit Card   
4                     439.92   2292.651500  1852.731500       Credit Card   
...                      ...           ...          ...               ...   
54995                 110.89   2819.907300  2709.017300        Debit Card   
54996                 488.59   3891.303900  3402.713900  Cash on Delivery   
54997                 417.63   4094.687500  3677.057500       Credit Card   
54998                 204.56   3977.473500  3772.913500       Credit Card   
54999                 380.81   3703.581900  3322.771900       Credit Card   

        Location  Discount Percentage  
0      Ahmedabad             8.865248  
1      Bangalore             3.776467  
2          Delhi            10.649563  
3          Delhi             0.000000  
4          Delhi            19.188263  
...          ...                  ...  
54995      Other             3.932399  
54996  Bangalore            12.555946  
54997     Mumbai            10.199313  
54998    Kolkata             5.142963  
54999     Mumbai            10.282208  

[55000 rows x 14 columns]

In [14]:
df["Discount Name"].isna().sum()

np.int64(0)

### 36. Drop the Discount Availed column.

In [29]:
df = df.drop("Discount Availed",axis = 1)

| Axis     | Meaning                       | Example      |
| -------- | ----------------------------- | ------------ |
| `axis=0` | Operate on **rows** (default) | Drop rows    |
| `axis=1` | Operate on **columns**        | Drop columns |


In [ ]:
# df.drop(5, axis=0)
# # axis=0 → down the rows
# # axis=1 → across columns


In [17]:
df.columns

Index(['CID', 'TID', 'Gender', 'Age Group', 'Purchase Date',
       'Product Category', 'Discount Name', 'Discount Amount (INR)',
       'Gross Amount', 'Net Amount', 'Purchase Method', 'Location',
       'Discount Percentage'],
      dtype='object')

### 37. Rename Purchase Method column to PaymentMethod.

In [27]:
df = df.rename(columns={"Purchase Method": "PaymentMethod"})

In [30]:
df.columns

Index(['CID', 'TID', 'Gender', 'Age Group', 'Purchase Date',
       'Product Category', 'Discount Name', 'Discount Amount (INR)',
       'Gross Amount', 'Net Amount', 'PaymentMethod', 'Location'],
      dtype='object')

In [ ]:
#df.rename(columns={'old': 'new'}, inplace=True)
## Original df is modified directly

### 38. Reorder columns so CID and TID come first.

In [ ]:
cols = ["CID","TID"] + [c for c in df.columns if c not in ['CID', 'TID']]   ## Doubt

In [32]:
df = df[cols]

In [34]:
df.columns

Index(['CID', 'TID', 'Gender', 'Age Group', 'Purchase Date',
       'Product Category', 'Discount Name', 'Discount Amount (INR)',
       'Gross Amount', 'Net Amount', 'PaymentMethod', 'Location'],
      dtype='object')

### 39. Check if TID column has duplicates.

In [36]:
df["TID"].duplicated().sum()

np.int64(0)

### 40. Drop duplicate rows based on TID.

In [37]:
df = df.drop_duplicates(subset=["TID"])

### 41. Convert Purchase Date to datetime format.

In [89]:
df["Purchase Date"] = pd.to_datetime(df["Purchase Date"], format = '%d/%m/%Y %H:%M:%S', errors="coerce")

In [ ]:
#“errors='coerce' converts invalid datetime values to NaT instead of raising an exception, allowing safe data processing.”

In [51]:
df["Purchase Date"].dtype

dtype('<M8[ns]')

In [52]:
# <M8 → datetime64
# [ns] → nanosecond precision

In [49]:
df["Purchase Date"]

0       2023-08-30 20:27:08
1       2024-02-23 09:33:46
2       2022-03-06 09:09:50
3       2020-11-04 04:41:57
4       2022-05-31 17:00:32
                ...        
54995   2020-02-23 10:42:04
54996   2021-07-12 15:59:21
54997   2024-08-02 09:30:44
54998   2020-08-05 23:57:56
54999   2022-07-21 09:05:18
Name: Purchase Date, Length: 55000, dtype: datetime64[ns]

### 42. Convert CID to integer type (handle missing values first).

In [55]:
df["CID"].isna().sum()

np.int64(0)

In [54]:
df["CID"].dtype

dtype('int64')

In [57]:
df['CID'] = df['CID'].fillna(0).astype(float)   # ANS

In [58]:
df["CID"].dtype

dtype('float64')

### 43. Convert Discount Amount (INR) to float type.

In [56]:
df["Discount Amount (INR)"].dtype

dtype('float64')

In [59]:
df["Discount Amount (INR)"] = df["Discount Amount (INR)"].astype(int)

In [60]:
df["Discount Amount (INR)"].dtype

dtype('int64')

In [61]:
df["Discount Amount (INR)"]

0         64
1        175
2        211
3          0
4        439
        ... 
54995    110
54996    488
54997    417
54998    204
54999    380
Name: Discount Amount (INR), Length: 55000, dtype: int64

### 44. Convert Gross Amount and Net Amount to float type.

In [64]:
df[["Gross Amount","Net Amount"]].dtypes

Gross Amount    float64
Net Amount      float64
dtype: object

In [65]:
df = df.astype({
    "Gross Amount":"int",
    "Net Amount":"int"
})

In [66]:
df[["Gross Amount","Net Amount"]].dtypes

Gross Amount    int64
Net Amount      int64
dtype: object

### 45. Convert Gender column to category type.

In [67]:
df["Gender"].dtype

dtype('O')

In [68]:
df["Gender"] = df["Gender"].astype("category")

In [69]:
df["Gender"].dtype

CategoricalDtype(categories=['Female', 'Male', 'Other'], ordered=False, categories_dtype=object)

### 46. Convert Location column to category type.

In [77]:
# Run 47 first

In [80]:
df["Location"].dtype

dtype('O')

In [71]:
df["Location"] = df["Location"].astype("category")

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   CID                    55000 non-null  float64       
 1   TID                    55000 non-null  int64         
 2   Gender                 55000 non-null  category      
 3   Age Group              55000 non-null  object        
 4   Purchase Date          55000 non-null  datetime64[ns]
 5   Product Category       55000 non-null  object        
 6   Discount Availed       55000 non-null  object        
 7   Discount Name          27415 non-null  object        
 8   Discount Amount (INR)  55000 non-null  int64         
 9   Gross Amount           55000 non-null  int64         
 10  Net Amount             55000 non-null  int64         
 11  Purchase Method        55000 non-null  object        
 12  Location               55000 non-null  category      
dtypes

In [72]:
df["Location"].dtype

CategoricalDtype(categories=['Ahmedabad', 'Bangalore', 'Chennai', 'Dehradun', 'Delhi',
                  'Hyderabad', 'Jaipur', 'Kolkata', 'Lucknow', 'Mumbai',
                  'Other', 'Pune', 'Srinagar', 'Varanasi'],
, ordered=False, categories_dtype=object)

### 47. Check memory usage before and after converting categorical columns.

In [ ]:
# deep=True is a parameter used when checking memory usage in Pandas.
# It tells Pandas to calculate the **actual memory consumed

In [78]:
df.memory_usage(deep=True)

Index                        132
CID                       440000
TID                       440000
Gender                   2970358
Age Group                2997101
Purchase Date            3740000
Product Category         3328417
Discount Availed         2832415
Discount Name            2489840
Discount Amount (INR)     440000
Gross Amount              440000
Net Amount                440000
Purchase Method          3322398
Location                 3066957
dtype: int64

In [79]:
df.memory_usage(deep=True).sum()


np.int64(26947618)

In [81]:
df["Location"] = df["Location"].astype("category")

In [82]:
df.memory_usage(deep=True)

Index                        132
CID                       440000
TID                       440000
Gender                   2970358
Age Group                2997101
Purchase Date            3740000
Product Category         3328417
Discount Availed         2832415
Discount Name            2489840
Discount Amount (INR)     440000
Gross Amount              440000
Net Amount                440000
Purchase Method          3322398
Location                   56340
dtype: int64

In [85]:
df.memory_usage(deep=True).sum()

np.int64(23937001)

Location                 3066957   ---> before
Location                   56340   ---> After

Significant memory reduction
Categorical encoding is effective


### 48. Extract the year from Purchase Date into a new column.

In [90]:
df['Purchase Date'].dtype


dtype('<M8[ns]')

In [92]:
df["Year"] = df["Purchase Date"].dt.year

In [96]:
df[["Purchase Date","Year"]].head()

Purchase Date  Year
0 2023-08-30 20:27:08  2023
1 2024-02-23 09:33:46  2024
2 2022-03-06 09:09:50  2022
3 2020-11-04 04:41:57  2020
4 2022-05-31 17:00:32  2022

### 49. Extract the month from Purchase Date into a new column.

In [97]:
df["Month"] = df["Purchase Date"].dt.month

In [98]:
df[["Purchase Date", "Month"]].head()

Purchase Date  Month
0 2023-08-30 20:27:08      8
1 2024-02-23 09:33:46      2
2 2022-03-06 09:09:50      3
3 2020-11-04 04:41:57     11
4 2022-05-31 17:00:32      5

### 50. Extract the day name from Purchase Date into a new column.

In [106]:
df["Week_Day"] = df["Purchase Date"].dt.day_name()

In [107]:
df[["Purchase Date","Week_Day"]].head()

Purchase Date   Week_Day
0 2023-08-30 20:27:08  Wednesday
1 2024-02-23 09:33:46     Friday
2 2022-03-06 09:09:50     Sunday
3 2020-11-04 04:41:57  Wednesday
4 2022-05-31 17:00:32    Tuesday